packages/plugin-kafka/src/custom/admin-client.ts

```diff
@@ -0,0 +1,127 @@
+  private service: Microfleet
+  private client!: Producer
+  private kafka: KafkaFactory
+
+  public adminClient: IAdminClient
+  public defaultWaitParams: RetryParams
+
+  constructor(service: Microfleet, kafka: KafkaFactory) {
+    this.service = service
+    this.kafka = kafka
+    this.adminClient = this.createAdminClient()
+
+    this.defaultWaitParams = {
+      tries: 10,
+      interval: 100,
+      timeout: 5000,
+    }
+  }
+
+  public close(): void {
+    this.service.log.debug('closing admin client')
+    this.adminClient.disconnect()
+  }
+
+  public async createTopic(req: CreateTopicRequest): Promise<void> {
+    const client = req.client || await this.getClient()
+    const criteria: WaitCriteria = topic => topic ? true : false
+
+    await this.adminClient.createTopicAsync(req.topic)
+    await this.waitFor(client, req.topic.topic, criteria, req.params)
+  }
+
+  public async deleteTopic(req: DeleteTopicRequest): Promise<void> {
+    const client = req.client || await this.getClient()
+    const criteria: WaitCriteria = topic => topic === undefined
+
+    await this.adminClient.deleteTopicAsync(req.topic)
+    await this.waitFor(client, req.topic, criteria, req.params)
+  }
+
+  private createAdminClient(): IAdminClient {
+    // node-rdkafka Admin client not exported and available only throught `create` function
+    // see https://github.com/Blizzard/node-rdkafka/blob/master/lib/admin.js#L11
+    const { kafka } = this
+    const adminClient = AdminClient.create(kafka.rdKafkaConfig)
+
+    promisifyAll(adminClient)
+    return adminClient
+  }
+
+  private async getClient(): Promise<KafkaClientType> {
+    if (!this.client) {
+      const { kafka } = this
+
+      this.client = kafka.createClient(Producer, kafka.rdKafkaConfig, {})
+      kafka.attachClientLogger(this.client, { type: 'admin-producer' })
+
+      await this.client.connectAsync({
+        allTopics: true,
+      })
+    }
+    return this.client
+  }
+
+  private async getTopicFromMeta(client: KafkaClient, topicName: string): Promise<TopicMetadata> {
+    const meta = await client.getMetadataAsync({
+      allTopics: true
+    })
+    const [filtered] = filterTopic(meta, topicName)
+    return filtered
+  }
+
+  private async waitFor(client: KafkaClient, topicName: string, criteria: WaitCriteria, params?: RetryParams): Promise<TopicMetadata> {
+    const { tries, timeout, interval } = merge(params, this.defaultWaitParams)
+    let attempts = 0
+
+    const waitLoop = async (): Promise<TopicMetadata> => {
+      while (attempts < tries) {
+        attempts += 1
+        const filtered = await this.getTopicFromMeta(client, topicName)
+        if (criteria(filtered)) return filtered
+        await delay(interval)
+      }
+      throw new TopicWaitError(`topic '${topicName}' wait error`,params, { attempts })
+    }
+
+    const topic = await Promise.race([
+      waitLoop(),
+      delay(timeout).throw(new TopicWaitError(params, { attempts } )),
```

Comment:
```suggestion
      delay(timeout).then(() => throw new TopicWaitError(params, { attempts } )),
```

Reason for the change is that new Error() is a very expensive call due to stacktrace and it will be executed every time

Reply:
wait loop will not exit after Promise.race
You need to explicitly exit it

Reply:
fixed